### DISCUSSION TITLE
Can i run slint and another winit thread in main thread on macOS?

### DISCUSSION THREAD

**Discussion by LeeeSe** [slint-ui/slint#3657] [Q&A]
I have a taskbar app [MessAuto](https://github.com/LeeeSe/MessAuto), [tray-icon](https://github.com/tauri-apps/tray-icon) where the event loop can only happen in the main thread (on macOS), I implemented a hover window today via slint and when I tried to merge them together I found it almost impossible, can I have any solution?

**Comment by ogoffart** [comment]
I'm not quite sure of your architecture. Are you using another library than Slint that uses Winit?
The Slint's winit backend will let winit handle the event loop. And that event loop must be in the main thread, it is a limitation of the OS.

If you can't use the winit backend, you can also control the Slint event loop yourself from any thread using the `slint::platform` API,  but that may be overkill.

**Comment by LeeeSe** [comment]
My App uses [tauri-apps/tray-icon](https://github.com/tauri-apps/tray-icon), it doesn't provide a winit backend, but it needs winit to run.

if I want to add a slint window to my App, then two winit backends will conflict, so for now I can only wait for slint to implement the task tray icon, [tray-icon](https://github.com/tauri-apps/tray-icon) is open source software, maybe slint can consider to insert tray-icon into event_loop.

If slint doesn't have any plans to implement the tray-icon feature, then I'm willing to wait, because I really enjoy developing Native GUIs with slint, and I hate web apps 😛 .

**Reply by ogoffart** [reply]
Slint can use the winit backend, you should force it by disabling the  backend-qt feature that is enabled by default.
You can also force it at runtime by setting the SLINT_BACKEND environment variable. See https://slint.dev/releases/1.2.2/docs/slint/src/advanced/backends_and_renderers

There is also internal unstable API to access the winit backend: https://docs.rs/i-slint-backend-winit/latest/i_slint_backend_winit/

I haven't tried to used tray-icon with Slint, but i think other people have done it.

**Comment by LeeeSe** [comment]
I using another process for my needs finally
```rust
use std::process::Command;

mod tray_icon;
mod app;

pub const ARGS_APP:&str = "app";

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() > 1{
        let arg1 = args[1].to_lowercase();
        if arg1.starts_with(ARGS_APP) {
            return app::main().unwrap();
        }
    }
    open_app();
    tray_icon::main();
}

pub fn open_app(){
    let _ = start_process(vec![ARGS_APP.to_string()]);
}

fn start_process(command_args: Vec<String>) {
    let current_exe = std::env::current_exe().unwrap();

    Command::new(current_exe)
        .args(&command_args)
        .spawn().unwrap();
}

```

edit
[from here](https://github.com/planet0104/slint_multi_window.git)
